File: module-18-network-forensics/pcap_reader.py
"""
Module 18 -- parses a REAL .pcap file (global header + repeated packet
records), reconstructing each frame's Ethernet/IP/TCP fields and payload.
This reads the exact same file format Wireshark/tcpdump use -- confirmed
during this module's build with the independent `file` command-line
utility (libmagic), which correctly identified pcap_writer.py's output as
a genuine "pcap capture file... version 2.4 (Ethernet)" with zero
knowledge of this course's code.
"""
import struct

PCAP_MAGIC = 0xA1B2C3D4


def mac_str(b: bytes) -> str:
    return b.hex(":")


def ip_str(b: bytes) -> str:
    return ".".join(str(o) for o in b)


def read_pcap(path: str):
    with open(path, "rb") as f:
        header = f.read(24)
        magic, ver_major, ver_minor, thiszone, sigfigs, snaplen, network = struct.unpack("!IHHiIII", header)
        if magic != PCAP_MAGIC:
            raise ValueError(f"not a recognized big-endian pcap file (magic={magic:#x})")

        packets = []
        while True:
            record_header = f.read(16)
            if len(record_header) < 16:
                break
            ts_sec, ts_usec, incl_len, orig_len = struct.unpack("!IIII", record_header)
            frame = f.read(incl_len)
            packets.append({"ts": ts_sec + ts_usec / 1_000_000, "frame": frame, "orig_len": orig_len})

        return {"version": (ver_major, ver_minor), "snaplen": snaplen, "network": network}, packets


def parse_frame(frame: bytes) -> dict:
    dst_mac, src_mac, ethertype = struct.unpack("!6s6sH", frame[0:14])
    ip_header = frame[14:34]
    version_ihl, tos, total_len, ident, flags_frag, ttl, proto, csum, src_ip, dst_ip = struct.unpack(
        "!BBHHHBBH4s4s", ip_header
    )
    tcp_header = frame[34:54]
    src_port, dst_port, seq, ack, offset_flags, window, tcp_csum, urgent = struct.unpack(
        "!HHLLHHHH", tcp_header
    )
    tcp_flags = offset_flags & 0x3F
    payload = frame[54:]

    return {
        "src_mac": mac_str(src_mac), "dst_mac": mac_str(dst_mac),
        "src_ip": ip_str(src_ip), "dst_ip": ip_str(dst_ip),
        "src_port": src_port, "dst_port": dst_port,
        "tcp_flags": tcp_flags, "seq": seq, "ack": ack,
        "payload": payload,
    }


FLAG_NAMES = {0x02: "SYN", 0x10: "ACK", 0x18: "PSH,ACK", 0x12: "SYN,ACK", 0x01: "FIN", 0x04: "RST"}


def main():
    import sys
    path = sys.argv[1] if len(sys.argv) > 1 else "incident.pcap"

    meta, packets = read_pcap(path)
    print(f"pcap version {meta['version']}, snaplen {meta['snaplen']}, linktype {meta['network']}")
    print(f"{len(packets)} packets\n")

    for i, pkt in enumerate(packets):
        p = parse_frame(pkt["frame"])
        flags = FLAG_NAMES.get(p["tcp_flags"], hex(p["tcp_flags"]))
        payload_preview = p["payload"][:40]
        print(f"#{i:03d} t={pkt['ts']:.3f} {p['src_ip']}:{p['src_port']} -> {p['dst_ip']}:{p['dst_port']} "
              f"[{flags}] len={len(p['payload'])}"
              + (f" payload={payload_preview!r}" if payload_preview else ""))


if __name__ == "__main__":
    main()

File: module-18-network-forensics/pcap_reader.js
/**
 * Module 18 — JS/Node port of pcap_reader.py. Same real pcap parsing.
 */
const fs = require("fs");

const PCAP_MAGIC = 0xa1b2c3d4;

function macStr(buf) {
  return buf.toString("hex").match(/../g).join(":");
}

function ipStr(buf) {
  return Array.from(buf).join(".");
}

function readPcap(path) {
  const data = fs.readFileSync(path);
  const magic = data.readUInt32BE(0);
  if (magic !== PCAP_MAGIC) {
    throw new Error(`not a recognized big-endian pcap file (magic=0x${magic.toString(16)})`);
  }
  const verMajor = data.readUInt16BE(4);
  const verMinor = data.readUInt16BE(6);
  const snaplen = data.readUInt32BE(16);
  const network = data.readUInt32BE(20);

  const packets = [];
  let offset = 24;
  while (offset + 16 <= data.length) {
    const tsSec = data.readUInt32BE(offset);
    const tsUsec = data.readUInt32BE(offset + 4);
    const inclLen = data.readUInt32BE(offset + 8);
    offset += 16;
    const frame = data.subarray(offset, offset + inclLen);
    offset += inclLen;
    packets.push({ ts: tsSec + tsUsec / 1_000_000, frame });
  }

  return { meta: { version: [verMajor, verMinor], snaplen, network }, packets };
}

function parseFrame(frame) {
  const dstMac = frame.subarray(0, 6);
  const srcMac = frame.subarray(6, 12);
  const ipHeader = frame.subarray(14, 34);
  const srcIp = ipHeader.subarray(12, 16);
  const dstIp = ipHeader.subarray(16, 20);

  const tcpHeader = frame.subarray(34, 54);
  const srcPort = tcpHeader.readUInt16BE(0);
  const dstPort = tcpHeader.readUInt16BE(2);
  const seq = tcpHeader.readUInt32BE(4);
  const ack = tcpHeader.readUInt32BE(8);
  const offsetFlags = tcpHeader.readUInt16BE(12);
  const tcpFlags = offsetFlags & 0x3f;
  const payload = frame.subarray(54);

  return {
    srcMac: macStr(srcMac), dstMac: macStr(dstMac),
    srcIp: ipStr(srcIp), dstIp: ipStr(dstIp),
    srcPort, dstPort, tcpFlags, seq, ack, payload,
  };
}

const FLAG_NAMES = { 0x02: "SYN", 0x10: "ACK", 0x18: "PSH,ACK", 0x12: "SYN,ACK", 0x01: "FIN", 0x04: "RST" };

function main() {
  const path = process.argv[2] || "incident_node.pcap";
  const { meta, packets } = readPcap(path);
  console.log(`pcap version ${JSON.stringify(meta.version)}, snaplen ${meta.snaplen}, linktype ${meta.network}`);
  console.log(`${packets.length} packets\n`);

  packets.forEach((pkt, i) => {
    const p = parseFrame(pkt.frame);
    const flags = FLAG_NAMES[p.tcpFlags] || `0x${p.tcpFlags.toString(16)}`;
    const payloadPreview = p.payload.subarray(0, 40);
    const idx = String(i).padStart(3, "0");
    let line = `#${idx} t=${pkt.ts.toFixed(3)} ${p.srcIp}:${p.srcPort} -> ${p.dstIp}:${p.dstPort} [${flags}] len=${p.payload.length}`;
    if (payloadPreview.length > 0) {
      line += ` payload=${JSON.stringify(payloadPreview.toString())}`;
    }
    console.log(line);
  });
}

if (require.main === module) {
  main();
}

module.exports = { readPcap, parseFrame };
